dali/operators/reader/loader/loader.h

```diff
@@ -213,14 +212,7 @@ class Loader {
   virtual void ReadSample(LoadTarget& tensor) = 0;
 
   void PrepareMetadata() {
-    if (!loading_flag_) {
-      std::lock_guard<std::mutex> l(prepare_metadata_mutex_);
-      if (!loading_flag_) {
-        PrepareMetadataImpl();
-        std::atomic_thread_fence(std::memory_order_release);
-        loading_flag_ = true;
-      }
-    }
+    std::call_once(loading_flag_, &Loader::PrepareMetadataImpl, this);
```

Comment:
It might be elegant, but the glibc implementation of call_once is
a) bugged - it fails when used before threading library is initialized
b) slow
This function is invoked for each sample. The check-lock-check idiom is 4x faster - but what's even more damning, is the fact that call_once uses atomics at the first line - this is not necessary and it causes a system-wide bus lock to be issued. That's an overkill on multi-socket systems which we target.
I've written a toy example where these two methods are used a million times and a background thread is incrementing a volatile (but not atomic) variable. The background thread slows down around 12x due to bus locks when using call_once, but is not affected by double-checking a boolean.

Reply:
The old code had used  ```std::atomic_thread_fence(std::memory_order_release);```  Should we use ```         std::atomic_thread_fence(std::memory_order_acquire );``` before first checking loading_flag_?